CODING_GUIDELINES.md

```diff
@@ -0,0 +1,690 @@
+It means that if an object is correctly designed, the resources associated with it can not survive its lifetime.
+Since objects that are allocated on the stack (local variables in a function, for example) are automatically destroyed when a function is exited, be it by reaching the last line, calling return or throwing an exception, it makes it possible to ensure that resources are always properly destroyed by wrapping them into an object.
+
+We describe the use of smart pointers, containers and other wrappers for that purpose below, but first a few words of caution.
+Resources stored in a object are only tied to this object if the constructor finished properly.
+If an exception is raised in the constructor body, the object is not created and therefore the destructor will not get called.
+This means that if the object has non-object members holding resources, like naked file descriptors or naked pointers, they need to be explicitly released before raising the exception, otherwise they are lost.
+
+```C++
+class BadFileDescriptorWrapper
+{
+  BadFileDescriptorWrapper()
+  {
+    d_fd = open(...);
+    if (something) {
+      throw std::runtime_error(...);
+    }
+    ...
+  }
+  ~BadFileDescriptorWrapper()
+  {
+    if (d_fd > 0) {
+      close(d_fd);
+      d_fd = -1;
+    }
+  }
+  int getHandle() const
+  {
+    return d_fd;
+  }
+private:
+  int d_fd{-1}; 
+};
+```
+
+The use of smart pointers can be a solution to most resources leakage, but otherwise the only way is to be careful about exceptions in constructors:
+
+```C++
+BadFileDescriptorWrapper()
+{
+  d_fd = open(...);
+  if (something) {
+    close(d_fd);
+    throw std::runtime_error(...);
+  }
+  ...
+}
+```
+
+## Smart pointers
+
+There is almost no good reason not to use a smart pointer when doing dynamic memory allocation.
+Smart pointers will keep track of whether the dynamically allocated object is still used, and destroy it when the last user goes away.
+
+Using naked pointers quickly results in security issues, ranging from memory leaks to arbitrary code execution.
+Examples of such issues can be found in the following PowerDNS security advisories:
+
+* [2017-07: Memory leak in DNSSEC parsing](https://docs.powerdns.com/recursor/security-advisories/powerdns-advisory-2017-07.html)
+* [2018-04: Crafted answer can cause a denial of service](https://docs.powerdns.com/recursor/security-advisories/powerdns-advisory-2018-04.html)
+
+Most allocations should be wrapped in a `std::unique_ptr`, using `make_unique`.
+There can only be one owner at a given time, as opposed to shared pointers, but the ownership can be passed along using `std::move()` if needed.
+
+If the dynamically allocated object needs to be referenced in several places, the use of a `std::shared_ptr` is advised instead, via `std::make_shared`.
+
+The use of the `make_*` methods has three advantages:
+
+* they result in a single allocation for `shared_ptr`s, instead of two otherwise ;
```

Comment:
I see ` ;` after a lot of list items, but not after all of them. Is this some leftover from the conversion from another format? Whatever we do, we should do it consistently.

Reply:
Will do!